crates/subspace-farmer/src/farmer_cache.rs

```diff
@@ -89,9 +90,35 @@ impl CacheBackend {
     fn new(backend: Arc<dyn PieceCache>, total_capacity: u32) -> Self {
         Self {
             backend,
+            used_capacity: 0,
             total_capacity,
         }
     }
+
+    fn next_free(&mut self) -> Option<PieceCacheOffset> {
+        let offset = self.used_capacity;
+        if offset < self.total_capacity {
+            self.used_capacity += 1;
+            Some(PieceCacheOffset(offset))
+        } else {
+            debug!(?offset, total_capacity = ?self.total_capacity, "No free space in cache backend");
+            None
+        }
+    }
+
+    fn free_size(&self) -> u32 {
+        self.total_capacity - self.used_capacity
+    }
+
+    fn dangling_free_offsets<CacheIndex>(
+        &self,
+        free_offsets: Vec<FarmerCacheOffset<CacheIndex>>,
+    ) -> Vec<FarmerCacheOffset<CacheIndex>> {
+        free_offsets
+            .into_iter()
+            .filter(|free_offset| free_offset.piece_offset.0 < self.used_capacity)
+            .collect()
+    }
```

Comment:
This method doesn't belong here. It doesn't use `self` and doesn't even use data structures of `CacheBackend`. In fact it is called in a single place, I'd inline the necessary logic there.

Reply:
Yes, its actually belongs to CacheState, Here's where the dangling_offsets are calculated